---
title: "Differential Binding at Peaks"
output:
  html_document:
       keep_md: true
  html_notebook: default
---


# Get sequencing file and align to dm6
Download sra-file and convert to fastq-file.
```{r, eval=FALSE}
module load ngs/sratoolkit/2.8.0

prefetch SRR520444 #Input_Kc_ctrl_1
prefetch SRR520445 #CLAMP_Kc_ctrl_1
prefetch SRR520446 #Input_Kc_ctrl_2
prefetch SRR520447 #CLAMP_Kc_ctrl_2
prefetch SRR520448 #Input_S2_msl2_1
prefetch SRR520449 #CLAMP_S2_msl2_1
prefetch SRR520450 #Input_S2_msl2_2
prefetch SRR520451 #CLAMP_S2_msl2_2
prefetch SRR520452 #Input_S2_ctrl_1
prefetch SRR520453 #CLAMP_S2_ctrl_1
prefetch SRR520454 #Input_S2_ctrl_2
prefetch SRR520455 #CLAMP_S2_ctrl_2

fastq-dump SRR520444 #Input_Kc_ctrl_1
fastq-dump SRR520445 #CLAMP_Kc_ctrl_1
fastq-dump SRR520446 #Input_Kc_ctrl_2
fastq-dump SRR520447 #CLAMP_Kc_ctrl_2
fastq-dump SRR520448 #Input_S2_msl2_1
fastq-dump SRR520449 #CLAMP_S2_msl2_1
fastq-dump SRR520450 #Input_S2_msl2_2
fastq-dump SRR520451 #CLAMP_S2_msl2_2
fastq-dump SRR520452 #Input_S2_ctrl_1
fastq-dump SRR520453 #CLAMP_S2_ctrl_1
fastq-dump SRR520454 #Input_S2_ctrl_2
fastq-dump SRR520455 #CLAMP_S2_ctrl_2
```

Align fastq-file with bowtie1 (with -m 1 option) to *Drosophila* genome dm6.
(here, reads mapped with quality score < 10 were removed, not necessary)
```{r, eval=FALSE}
module load ngs/bowtie1
module load ngs/samtools

bowtie -S -p 18 -m 1 /work/data/genomes/fly/Drosophila_melanogaster/UCSC/dm6/Sequence/BowtieIndex/genome ${FILENAME}.fastq | samtools view -bS -q 10 - | samtools sort - | tee ${FILENAME}.sorted.bam | samtools index - ${FILENAME}.sorted.bam.bai
```

# Create count-file from bam-file - BAM2count.R
Read in bam-files, convert to GRanged Object, extend read to 200 bp and save as .rda-file using the BAM2count.R skript.
(Usage: BAM2count.R <bam file> <fragment.length>,
Example: BAM2count.R input_gst_1 200)
```{r, eval=FALSE}
library(IRanges)
library(ShortRead)
library(rtracklayer)
library(GenomicAlignments)

srbam2count <- function(bam.file, extend=200) {
  require(ShortRead)
  gr <- readGAlignments(bam.file)
  grs <- as(gr, "GRanges")
  grsr <- resize(grs, extend)
  return(grsr)
}

args <- commandArgs(trailingOnly = TRUE)
file <- args[[1]]
fragment.length <- as.integer(args[[2]])

counts <- srbam2count(file, extend=fragment.length)
save(counts, file=paste(file,".counts.rda",sep=""))
```

# Compute log2(IP/input) enrichment for peaks
Load required packages and load peak file (bed format), seqnames must be changed (e.g. '2R' > 'chr2R') to match count file seqnames; and convert to GRanged Object.
```{r, warning=F, error=F}
options(scipen=100)
rm(list=ls())

suppressPackageStartupMessages({
library(rtracklayer)
library(limma)
library(RColorBrewer)
})

HAS <- read.delim("msl2_in_vivo_peaks.bed", header = F)
HAS[,1] <- paste0('chr',HAS[,1])
colnames(HAS) <- c('chr','start','end')
HAS <- as(HAS, 'GRanges')

kc_peaks <- import.bed('CLAMP_Kc_ctrl_robust_peaks.bed')
s2_peaks <- import.bed('CLAMP_S2_ctrl_robust_peaks.bed')

overlaps <- queryHits(findOverlaps(s2_peaks,kc_peaks))
s2_peaks$kc <- 'n'
s2_peaks$kc[overlaps] <- 'y'

overlaps <- queryHits(findOverlaps(s2_peaks,HAS))
s2_peaks$has <- 'n'
s2_peaks$has[overlaps] <- 'y'

clamp_peaks <- subset(s2_peaks, kc == 'y' | has == 'y')
```

Load count files in for loop and count overlapping reads at peaks with minimal overlap of 100 bp (half read length) and store overall number of aligned reads.
```{r, eval=FALSE}
files <- system("ls *counts.rda",intern=T)
names <- unlist(strsplit(files,split = "_counts.rda"))[seq(1,length(files)*2,2)]

for(i in seq_along(files)){
  load(files[i])
  if(i==1){
    counts <- countOverlaps(clamp_peaks, counts, minoverlap = 100, ignore.strand=T)
    reads <- length(counts)
  }else{
    counts <- cbind(counts,countOverlaps(clamp_peaks, counts, minoverlap = 100, ignore.strand=T))
    reads <- c(reads,length(counts))
  }
}

colnames(counts) <- names
saveRDS(counts, "clamp_counts.rds")
names(reads) <- names
saveRDS(reads, "clamp_reads.rds")
```

Calculate size factor to acount for different sequencing depth. Size Factor is calcuted as propoartin of number of reads for each sample to the mean read number.
```{r}
counts <- readRDS("clamp_counts.rds")
head(counts)
reads <- readRDS("clamp_reads.rds")
reads
sizeFactor <- reads/mean(reads)
sizeFactor
counts.norm <- t(t(counts) / sizeFactor)
head(counts.norm)
zeros <- which(apply(counts.norm,1,function(x){min(x) == 0}))
counts.norm <- counts.norm[-zeros,]
clamp_peaks <- clamp_peaks[-zeros,]
```

Calculate log2(IP/input) and plot log2 fold change *mle* RNAi - *gst* RNAi.
```{r, fig.width=10, fig.height=5}
enrichment <- log2(counts.norm[,1:6]/counts.norm[,7:12])
head(enrichment)
S2.ctrl <- enrichment[,3:4]
S2.msl2 <- enrichment[,5:6]
Kc_ctrl <- enrichment[,1:2]

my_color <- brewer.pal(9, 'Set1')

par(mfrow = c(1,2))
plot(rowMeans(S2.ctrl[clamp_peaks$has == 'n',]),
     rowMeans(Kc_ctrl[clamp_peaks$has == 'n',]),
     main = "CLAMP enrichment at peaks",
     ylab = "Kc", xlab = "S2",
     pch=19, col=paste0(my_color[9],'5F'),
     ylim = c(-2, 8), xlim = c(-2,8))
points(rowMeans(S2.ctrl[clamp_peaks$has == 'y',]),
     rowMeans(Kc_ctrl[clamp_peaks$has == 'y',]),
     pch=19, col=paste0(my_color[1],'5F'))

plot(rowMeans(S2.ctrl[clamp_peaks$has == 'n',]),
     rowMeans(S2.msl2[clamp_peaks$has == 'n',]),
     main = "CLAMP enrichment at peaks",
     ylab = "msl2 RNAi S2", xlab = "ctrl RNAi S2",
     pch=19, col=paste0(my_color[9],'5F'),
     ylim = c(-2, 8), xlim = c(-2,8))
points(rowMeans(S2.ctrl[clamp_peaks$has == 'y',]),
     rowMeans(S2.msl2[clamp_peaks$has == 'y',]),
     pch=19, col=paste0(my_color[1],'5F'))
```

# Test for significantly differnt log2(IP/input) enrichment values at peaks
To test for significant differences use the limma package.
Create 'Design Matrix' for fitting a linear model with lmFit() and calculate statistices with eBayes() allow to use intensity-trend for prior variance and allow the prior estimation to be robustified against outlier sample variance. Get ranked table with topTable() using 'fdr method' to adjust for multiple testing.
```{r}
my_RNAi <- c(rep("gst",2),rep("msl2",2))
my_RNAi <- factor(my_RNAi, levels = c("gst", "msl2"))
design <- model.matrix(~ my_RNAi)
design

fit <- lmFit(enrichment[,3:6], design)
fit <- eBayes(fit, trend=T, robust=T)
tt <- topTable(fit, coef=ncol(design), adjust.method = 'fdr', number = length(clamp_peaks))
head(tt)
```


```{r, fig.width=5, fig.height=5}
fdr <- seq(0,1,0.001)
roc_mat <- matrix(0,nrow = length(fdr), ncol = 3)
colnames(roc_mat) <- c("tpr","fpr","fdr")
for(k in seq_along(fdr)){
  clamp_peaks$test <- "y"
  clamp_peaks$test[as.numeric(rownames(subset(tt, adj.P.Val < fdr[k])))] <- 'n'
  roc_mat[k,"tpr"] <- sum(clamp_peaks$kc == 'n' & clamp_peaks$test == 'n')/sum(clamp_peaks$kc == 'n')
    roc_mat[k,"fpr"] <- sum(clamp_peaks$kc == 'y' & clamp_peaks$test == 'n')/sum(clamp_peaks$kc == 'y')
    roc_mat[k,"fdr"] <- fdr[k]
}
roc_mat <- as.data.frame(roc_mat)
plot(tpr ~ fpr, roc_mat, type="l",lwd=3,col=my_color[1],
     ylab = 'true positive rate', xlab = 'false positive rate', main = 'ROC')
abline(0,1, lwd=2)
points(tpr ~ fpr, subset(roc_mat, fdr %in% c(0.01, 0.05, 0.1, 0.5)), pch = 16)

```

Plot log2 fold change *msl2* RNAi - *ctrl* RNAi and highlight significantly different peaks (FDR < 0.1 and < 0.5).
```{r, fig.width=10, fig.height=5}

fdr0.5 <- clamp_peaks[as.numeric(rownames(subset(tt, adj.P.Val < 0.5)))]
fdr0.1 <- clamp_peaks[as.numeric(rownames(subset(tt, adj.P.Val < 0.1)))]

par(mfrow = c(1,2))
plot(rowMeans(S2.ctrl)[as.numeric(rownames(subset(tt, adj.P.Val >= 0.5)))],
     rowMeans(S2.msl2-S2.ctrl)[as.numeric(rownames(subset(tt, adj.P.Val >= 0.5)))],
     main = "CLAMP enrichment at peaks",
     ylab = "lfc (msl2 RNAi - ctrl RNAi)", xlab = "CLAMP enrichment",
     pch=19, col=paste0(my_color[9],'8F'), 
     ylim = c(-6, 6), xlim = c(0,8))
points(rowMeans(S2.ctrl)[as.numeric(rownames(subset(tt, adj.P.Val < 0.5)))],
       rowMeans(S2.msl2-S2.ctrl)[as.numeric(rownames(subset(tt, adj.P.Val < 0.5)))],
       pch=19, col = paste0(my_color[1],'8F'))
legend('bottomleft', bty = 'n', legend = paste0('fdr < 0.5, n=',nrow(subset(tt, adj.P.Val < 0.5))),
       pch = 19, col = my_color[1])

plot(rowMeans(S2.ctrl)[as.numeric(rownames(subset(tt, adj.P.Val >= 0.1)))],
     rowMeans(S2.msl2-S2.ctrl)[as.numeric(rownames(subset(tt, adj.P.Val >= 0.1)))],
     main = "CLAMP enrichment at peaks",
     ylab = "lfc (msl2 RNAi - ctrl RNAi)", xlab = "CLAMP enrichment",
     pch=19, col=paste0(my_color[9],'8F'), 
     ylim = c(-6, 6), xlim = c(0,8))
points(rowMeans(S2.ctrl)[as.numeric(rownames(subset(tt, adj.P.Val < 0.1)))],
       rowMeans(S2.msl2-S2.ctrl)[as.numeric(rownames(subset(tt, adj.P.Val < 0.1)))],
       pch=19, col = paste0(my_color[1],'8F'))
legend('bottomleft', bty = 'n', legend = paste0('fdr < 0.1, n=',nrow(subset(tt, adj.P.Val < 0.1))), pch = 19, col = my_color[1])

```


Convert count files to coverage vector and store in List.
```{r, eval=F}
suppressPackageStartupMessages({
library(ShortRead)
library(rtracklayer)
library(GenomicRanges)
})

files <- system("ls CLAMP*counts.rda",intern=T)
names <- unlist(strsplit(files,split = ".fastq"))[seq(1,length(files)*2,2)]

covList <- list()
for(i in seq_along(files)){
  load(files[i])
  covList[[i]] <- coverage(counts.ip)/length(counts.ip)*10^6}
names(covList) <- names
saveRDS(covList, "clamp_covList.rds")
```

Plot Browser profile at region of interest using "tsTools"
```{r,fig.width=9, fig.height=7, warning=F, error=F}
suppressPackageStartupMessages({
  library(ShortRead)
  library(rtracklayer)
  library(GenomicRanges)
  library(gdata)
  library(GenomicFeatures)
  library(tsTools)
  library(RColorBrewer)
})

covList <- readRDS("clamp_covList.rds")

txdb <- makeTxDbFromGFF("~/Desktop/mount/work/data/genomes/fly/Drosophila_melanogaster/UCSC/dm6/Annotation/Archives/archive-2015-07-24-09-25-49/Genes/genes.gtf")

cols <- c(brewer.pal(4, "Greens")[3:4],brewer.pal(4, "Blues")[3:4],brewer.pal(4, "Reds")[3:4])
lim <- c(0,100)
lims <- rep(list(lim),6)

chr <- "chrX"
start <- 10*10^6
end <- 11*10^6

plotProfiles(start,end,chr,profs=covList,txdb=txdb,collapse=T, plot.labels=F, cols=cols, grid=T, ylims=lims,ylab="coverage (rpm)")
```

```{r, fig.width=9, fig.height=3, warning=F, error=F}
peakList <- list()
peakList[[1]] <- coverage(subset(clamp_peaks, has == "y"))
peakList[[2]] <- coverage(subset(clamp_peaks, has == "y" & kc == "n"))
peakList[[3]] <- coverage(subset(fdr0.5, has == "y"))
peakList[[4]] <- coverage(subset(fdr0.1, has == "y"))
names(peakList) <- c('HAS','non-Kc','FDR < 0.5','FDR < 0.1')

cols <- brewer.pal(4, "Set1")
lim <- c(0,1)
lims <- rep(list(lim),5)

chr <- "chrX"
start <- 10*10^6
end <- 11*10^6

plotProfiles(start,end,chr,profs=peakList,txdb=txdb,collapse=T, plot.labels=F, cols=cols, grid=T, ylims=lims,ylab="peaks")
```

