---
title: "Homer"
output: 
    html_document:
        keep_md: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(cache=TRUE)
```


## Peak Finding

```{bash, eval=FALSE}
#!/bin/sh
# homer.sh


# make pooled tag directory for bam files

makeTagDirectory E03pool_K36.dir ../BAM/E03A_K36_TTAGGCAT_1.bam ../BAM/E03B_K36_ACAGTGAT_1.bam
makeTagDirectory E11pool_K36.dir ../BAM/E11A_K36_GCCAATAT_1.bam ../BAM/E11B_K36_GATCAGAT_1.bam
makeTagDirectory E03pool_INP.dir ../BAM/E03A_INP_ATCACGAT_1.bam ../BAM/E03B_INP_CGATGTAT_1.bam
makeTagDirectory E11pool_INP.dir ../BAM/E11A_INP_TTAGGCAT_1.bam ../BAM/E11B_INP_ACAGTGAT_1.bam


# peak finding against input

findPeaks E03pool_K36.dir/  -i E03pool_INP.dir/ -style histone -C 0 -F 2 -o E03pool_K36.histone.F2.txt
findPeaks E11pool_K36.dir/  -i E11pool_INP.dir/ -style histone -C 0 -F 2 -o E11pool_K36.histone.F2.txt

findPeaks E03pool_K36.dir/  -i E03pool_INP.dir/ -style histone -C 0 -F 3 -o E03pool_K36.histone.F3.txt
findPeaks E11pool_K36.dir/  -i E11pool_INP.dir/ -style histone -C 0 -F 3 -o E11pool_K36.histone.F3.txt

findPeaks E03pool_K36.dir/  -i E03pool_INP.dir/ -region -size 500 -minDist 100 -C 0 -F 2 -o E03pool_K36.size500.minDist100.F2.txt
findPeaks E11pool_K36.dir/  -i E11pool_INP.dir/ -region -size 200 -minDist 100 -C 0 -F 2 -o E11pool_K36.size500.minDist100.F2.txt

findPeaks E03pool_K36.dir/  -i E03pool_INP.dir/ -region -size 500 -minDist 500 -C 0 -F 2 -o E03pool_K36.size500.minDist500.F2.txt
findPeaks E11pool_K36.dir/  -i E11pool_INP.dir/ -region -size 200 -minDist 500 -C 0 -F 2 -o E11pool_K36.size500.minDist500.F2.txt



# convert txt to BED

cat E03pool_K36.histone.F2.txt | grep -v "#" | cut -f 2,3,4 > E03pool_K36.histoneF2.bed
cat E11pool_K36.histone.F2.txt | grep -v "#" | cut -f 2,3,4 > E11pool_K36.histoneF2.bed

cat E03pool_K36.histone.F3.txt | grep -v "#" | cut -f 2,3,4 > E03pool_K36.histoneF3.bed
cat E11pool_K36.histone.F3.txt | grep -v "#" | cut -f 2,3,4 > E11pool_K36.histoneF3.bed

cat E03pool_K36.size500.minDist100.F2.txt | grep -v "#" | cut -f 2,3,4 > E03pool_K36.size500.minDist100.F2.bed
cat E11pool_K36.size500.minDist100.F2.txt | grep -v "#" | cut -f 2,3,4 > E11pool_K36.size500.minDist100.F2.bed

cat E03pool_K36.size500.minDist500.F2.txt | grep -v "#" | cut -f 2,3,4 > E03pool_K36.size500.minDist500.F2.bed
cat E11pool_K36.size500.minDist500.F2.txt | grep -v "#" | cut -f 2,3,4 > E11pool_K36.size500.minDist500.F2.bed



# all pooled for -F 2

cat E03pool_K36.histoneF2.bed E11pool_K36.histoneF2.bed >  Eallpool_K36.histoneF2.bed
sort -k1,1 -k2,2n Eallpool_K36.histoneF2.bed > Eallpool_K36.histoneF2.sorted.bed
mergeBed -i Eallpool_K36.histoneF2.sorted.bed > Eallpool_K36.histoneF2.merged.bed

cat E03pool_K36.size500.minDist100.F2.bed E11pool_K36.size500.minDist100.F2.bed >  Eallpool_K36.size500.minDist100.F2.bed
sort -k1,1 -k2,2n Eallpool_K36.size500.minDist100.F2.bed > Eallpool_K36.size500.minDist100.F2.sorted.bed
mergeBed -i Eallpool_K36.size500.minDist100.F2.sorted.bed > Eallpool_K36.size500.minDist100.F2.merged.bed

cat E03pool_K36.size500.minDist500.F2.bed E11pool_K36.size500.minDist500.F2.bed >  Eallpool_K36.size500.minDist500.F2.bed
sort -k1,1 -k2,2n Eallpool_K36.size500.minDist500.F2.bed > Eallpool_K36.size500.minDist500.F2.sorted.bed
mergeBed -i Eallpool_K36.size500.minDist500.F2.sorted.bed > Eallpool_K36.size500.minDist500.F2.merged.bed
```



## BAM Files

```{r, message=FALSE, warning=FALSE}
library(GenomicAlignments)
library(GenomicRanges)
library(csaw)
library(edgeR)

my_ChIP <- "K36"
my_control <- "INP"

bam.files <- list.files("../BAM/", pattern = ".bam$")
bam.files <- bam.files[grep(paste(my_ChIP,my_control,sep="|"), bam.files)]
bam.files.path <- file.path("../BAM/", bam.files)

my_samples <- gsub("_[G,A,T,C].*","",bam.files)

print(bam.files)
print(my_samples)
```



## Counting

```{r, message=FALSE, warning=FALSE, echo=FALSE}
source("functions.R")
```


```{r, message=FALSE, warning=FALSE, fig.align="center" }

my_peaks <- read.delim("Eallpool_K36.size500.minDist500.F2.merged.bed", header = FALSE)
colnames(my_peaks) <- c("chr","start","end")

my_peak_ranges <- makeGRangesFromDataFrame(my_peaks)

my_peak_counts <- my_peak_ranges

for(i in seq_along( bam.files.path)){

    my_peak_counts <- bam2bins(bam.file =  bam.files.path[i],
                                 bins = my_peak_counts,
                                 my_sample = my_samples[i]
                                )
}


saveRDS(my_peak_counts, paste(my_ChIP,"counts.rds", sep="."))

my_peak_counts
```



## Trended Bias

```{r, message=FALSE, warning=FALSE, fig.align="center"}

mcols(my_peak_counts) <- mcols(my_peak_counts)[!(grepl("INP", colnames(mcols(my_peak_counts) )))]


par(mfrow=c(1,2), oma=c(3,0,0,0),mar=c(5,4,4,1), mgp = c(2.5,1,0),
    cex=1, cex.axis=1, cex.lab=1.25, cex.main=1.5)

win.ab <- aveLogCPM(DGEList(as.matrix(mcols(my_peak_counts))))
adjc <- log2(as.matrix(mcols(my_peak_counts))+0.5)
logfc <- adjc[,1] - adjc[,4]
smoothScatter(win.ab, logfc, ylim=c(-3, 3), xlim=c(2, 12),
              xlab="Average abundance", ylab="Log-fold change")


offsets <- normOffsets(as.matrix(mcols(my_peak_counts)), type="loess")
norm.adjc <- adjc - offsets/log(2)
norm.fc <- norm.adjc[,1]-norm.adjc[,4]
smoothScatter(win.ab, norm.fc, ylim=c(-3, 3), xlim=c(2, 12),
              xlab="Average abundance", ylab="Log-fold change")
```



## Estimate Variability

```{r, message=FALSE, warning=FALSE, fig.align="center"}
stage <- factor(gsub("[A-B]_.*","", bam.files[!(grepl("INP", bam.files))]))
design <- model.matrix(~0+stage)
colnames(design) <- levels(stage)
design

y <- DGEList(as.matrix(mcols(my_peak_counts)))
y <- scaleOffset(y, offsets)
y <- estimateDisp(y, design)
summary(y$trended.dispersion)

fit <- glmQLFit(y, design, robust=TRUE)

par(mfrow=c(1,2), oma=c(3,0,0,0),mar=c(5,4,4,1), mgp = c(2.5,1,0),
    cex=1, cex.axis=1, cex.lab=1.25, cex.main=1.5)

plotBCV(y)

plotQLDisp(fit)

summary(fit$df.prior)
```



## Differential Test

```{r, message=FALSE, warning=FALSE}
contrast <- makeContrasts(E11-E03, levels=design)
res <- glmQLFTest(fit, contrast=contrast)

toptags <- topTags(res, n = nrow(res$table), sort.by = "none" )

toptags[1:5,]
```


## Region Annotation

```{r, message=FALSE, warning=FALSE}
out.ranges <- granges(my_peak_counts, use.mcols=FALSE)

mcols(out.ranges) <- cbind(mcols(out.ranges), toptags$table)

out.ranges$direction <- ifelse(out.ranges$logFC > 0, "up", "down")

library(org.Dm.eg.db)
library(TxDb.Dmelanogaster.UCSC.dm6.ensGene)

my_genes <- genes(TxDb.Dmelanogaster.UCSC.dm6.ensGene)
my_genes <- my_genes[seqnames(my_genes) %in% unique(seqnames(out.ranges))]

out.ranges <- out.ranges[seqnames(out.ranges) %in% unique(seqnames(my_genes))]

mid.ranges <- makeGRangesFromDataFrame(data.frame(chr=seqnames(out.ranges),
                                                   start=mid(ranges(out.ranges)),
                                                   end=mid(ranges(out.ranges))))

my_nearest <- nearest(mid.ranges,my_genes)
out.ranges$gene_id <- my_genes[my_nearest]$gene_id

out.ranges$gene_name <- mapIds(org.Dm.eg.db, out.ranges$gene_id, "SYMBOL", keytype="ENSEMBL", multiVals="first")

saveRDS(out.ranges, paste(my_ChIP,"results.rds", sep="."))

out.ranges
```


